### PR TITLE
Fix fmad issue

### LINKED ISSUE #550: Gradient error in forward AD
Running the following code would return gradient error on the `dynamichmc-support` branch:

```julia

using Turing, LinearAlgebra

# Sample a precision matrix A from a Wishart distribution
# with identity scale matrix and 250 degrees of freedome
dim2 = 250
A   = rand(Wishart(dim2, Matrix{Float64}(I, dim2, dim2)))
d   = MvNormal(zeros(dim2), A)

# 250-dimensional multivariate normal (MVN)
@model mdemo(d, N) = begin
    Θ = Vector(undef, N)
   for n=1:N
      Θ[n] ~ d
   end
end


Turing.setadbackend(:forward_diff)
chain = sample(mdemo(d, 1), HMC(5000, 0.1, 5)) # takes 3 mins
```

Error message:

```julia
julia> chain = sample(mdemo(d, 1), HMC(5000, 0.1, 5)) # takes 3 mins
[ Info:  Assume - `Θ` is a parameter
┌ Warning: Requires now needs a UUID; please see the readme for changes in 0.7.
└ @ Turing ~/.julia/dev/Turing/src/samplers/support/adapt.jl:60
ERROR: MethodError: no method matching Float64(::ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10})
Closest candidates are:
  Float64(::Real, ::RoundingMode) where T<:AbstractFloat at rounding.jl:185
  Float64(::T<:Number) where T<:Number at boot.jl:725
  Float64(::Int8) at float.jl:60
  ...
Stacktrace:
 [1] convert(::Type{Float64}, ::ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10}) at ./number.jl:7
 [2] convert(::Type{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Float64,10}}, ::ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10}) at /Users/hg344/.julia/packages/ForwardDiff/hnKaN/src/dual.jl:318
 [3] setindex! at ./array.jl:769 [inlined]
 [4] copyto!(::IndexLinear, ::Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Float64,10},1}, ::IndexLinear, ::Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10},1}) at ./abstractarray.jl:731
 [5] copyto!(::Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Float64,10},1}, ::Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10},1}) at ./abstractarray.jl:723
 [6] \(::Cholesky{Float64,Array{Float64,2}}, ::Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10},1}) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.0/LinearAlgebra/src/factorization.jl:86
 [7] \ at /Users/hg344/.julia/packages/PDMats/mL7bX/src/pdmat.jl:40 [inlined]
 [8] invquad(::PDMats.PDMat{Float64,Array{Float64,2}}, ::Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10},1}) at /Users/hg344/.julia/packages/PDMats/mL7bX/src/pdmat.jl:67
 [9] sqmahal(::MvNormal{Float64,PDMats.PDMat{Float64,Array{Float64,2}},Array{Float64,1}}, ::Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10},1}) at /Users/hg344/.julia/packages/Distributions/WHjOk/src/multivariate/mvnormal.jl:263
 [10] _logpdf(::MvNormal{Float64,PDMats.PDMat{Float64,Array{Float64,2}},Array{Float64,1}}, ::Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10},1}) at /Users/hg344/.julia/packages/Distributions/WHjOk/src/multivariate/mvnormal.jl:113
 [11] logpdf at /Users/hg344/.julia/packages/Distributions/WHjOk/src/multivariates.jl:175 [inlined]
 [12] logpdf_with_trans(::MvNormal{Float64,PDMats.PDMat{Float64,Array{Float64,2}},Array{Float64,1}}, ::Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10},1}, ::Bool) at /Users/hg344/.julia/dev/Turing/src/utilities/transform.jl:320
 [13] macro expansion at ./logging.jl:312 [inlined]
 [14] assume(::Turing.Sampler{HMCDA{Any}}, ::MvNormal{Float64,PDMats.PDMat{Float64,Array{Float64,2}},Array{Float64,1}}, ::Turing.VarReplay.VarName, ::Turing.VarReplay.VarInfo) at /Users/hg344/.julia/dev/Turing/src/samplers/hmc.jl:214
 [15] macro expansion at /Users/hg344/.julia/dev/Turing/src/core/compiler.jl:91 [inlined]
 [16] #mdemo_model#4(::Int64, ::MvNormal{Float64,PDMats.PDMat{Float64,Array{Float64,2}},Array{Float64,1}}, ::Function, ::Turing.VarReplay.VarInfo, ::Turing.Sampler{HMCDA{Any}}) at /Users/hg344/.julia/packages/MacroTools/4AjBS/src/utils.jl:297
 [17] mdemo_model(::Turing.VarReplay.VarInfo, ::Turing.Sampler{HMCDA{Any}}) at /Users/hg344/.julia/packages/MacroTools/4AjBS/src/utils.jl:297
 [18] #invokelatest#1 at ./essentials.jl:686 [inlined]
 [19] invokelatest at ./essentials.jl:685 [inlined]
 [20] runmodel!(::Function, ::Turing.VarReplay.VarInfo, ::Turing.Sampler{HMCDA{Any}}) at /Users/hg344/.julia/dev/Turing/src/samplers/support/hmc_core.jl:65
 [21] (::getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}})(::Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10},1}) at /Users/hg344/.julia/dev/Turing/src/core/ad.jl:52
 [22] chunk_mode_gradient!(::Array{Real,1}, ::getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}}, ::SubArray{Real,1,Array{Real,1},Tuple{Array{Int64,1}},false}, ::ForwardDiff.GradientConfig{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10,Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10},1}}) at /Users/hg344/.julia/packages/ForwardDiff/hnKaN/src/gradient.jl:148
 [23] gradient! at /Users/hg344/.julia/packages/ForwardDiff/hnKaN/src/gradient.jl:37 [inlined]
 [24] gradient!(::Array{Real,1}, ::getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}}, ::SubArray{Real,1,Array{Real,1},Tuple{Array{Int64,1}},false}, ::ForwardDiff.GradientConfig{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10,Array{ForwardDiff.Dual{ForwardDiff.Tag{getfield(Turing, Symbol("#f#76")){Turing.VarReplay.VarInfo,typeof(mdemo_model),Turing.Sampler{HMCDA{Any}}},Real},Real,10},1}}) at /Users/hg344/.julia/packages/ForwardDiff/hnKaN/src/gradient.jl:33
 [25] gradient_forward(::SubArray{Real,1,Array{Real,1},Tuple{Array{Int64,1}},false}, ::Turing.VarReplay.VarInfo, ::Function, ::Turing.Sampler{HMCDA{Any}}, ::Int64) at /Users/hg344/.julia/dev/Turing/src/core/ad.jl:58
 [26] gradient_forward at /Users/hg344/.julia/dev/Turing/src/core/ad.jl:47 [inlined]
 [27] gradient at /Users/hg344/.julia/dev/Turing/src/core/ad.jl:21 [inlined]
 [28] gradient at /Users/hg344/.julia/dev/Turing/src/core/ad.jl:19 [inlined]
 [29] #48 at /Users/hg344/.julia/dev/Turing/src/samplers/support/hmc_core.jl:10 [inlined]
 [30] #_leapfrog#56(::getfield(Turing, Symbol("##52#53")){Turing.VarReplay.VarInfo,Turing.Sampler{HMCDA{Any}}}, ::getfield(Turing, Symbol("##54#55")){Turing.Sampler{HMCDA{Any}}}, ::Function, ::SubArray{Real,1,Array{Real,1},Tuple{Array{Int64,1}},false}, ::Array{Float64,1}, ::Int64, ::Float64, ::getfield(Turing, Symbol("##48#49")){Turing.VarReplay.VarInfo,Turing.Sampler{HMCDA{Any}},typeof(mdemo_model)}) at /Users/hg344/.julia/dev/Turing/src/samplers/support/hmc_core.jl:93
 [31] (::getfield(Turing, Symbol("#kw##_leapfrog")))(::NamedTuple{(:rev_func, :log_func),Tuple{getfield(Turing, Symbol("##52#53")){Turing.VarReplay.VarInfo,Turing.Sampler{HMCDA{Any}}},getfield(Turing, Symbol("##54#55")){Turing.Sampler{HMCDA{Any}}}}}, ::typeof(Turing._leapfrog), ::SubArray{Real,1,Array{Real,1},Tuple{Array{Int64,1}},false}, ::Array{Float64,1}, ::Int64, ::Float64, ::Function) at ./none:0
 [32] #_hmc_step#60(::Function, ::Function, ::Function, ::SubArray{Real,1,Array{Real,1},Tuple{Array{Int64,1}},false}, ::Float64, ::getfield(Turing, Symbol("##50#51")){Turing.VarReplay.VarInfo,Turing.Sampler{HMCDA{Any}},typeof(mdemo_model)}, ::Function, ::Float64, ::Float64, ::Array{Float64,1}) at ./logging.jl:312
 [33] (::getfield(Turing, Symbol("#kw##_hmc_step")))(::NamedTuple{(:rev_func, :log_func),Tuple{getfield(Turing, Symbol("##52#53")){Turing.VarReplay.VarInfo,Turing.Sampler{HMCDA{Any}}},getfield(Turing, Symbol("##54#55")){Turing.Sampler{HMCDA{Any}}}}}, ::typeof(Turing._hmc_step), ::SubArray{Real,1,Array{Real,1},Tuple{Array{Int64,1}},false}, ::Float64, ::Function, ::Function, ::Float64, ::Float64, ::Array{Float64,1}) at ./none:0
 [34] macro expansion at ./logging.jl:319 [inlined]
 [35] step(::Function, ::Turing.Sampler{HMCDA{Any}}, ::Turing.VarReplay.VarInfo, ::Bool) at /Users/hg344/.julia/dev/Turing/src/samplers/hmcda.jl:82
 [36] macro expansion at ./logging.jl:309 [inlined]
 [37] #sample#63(::Int64, ::Bool, ::Nothing, ::Int64, ::Stan.Adapt, ::Symbol, ::Function, ::Function, ::HMC{Any}) at /Users/hg344/.julia/dev/Turing/src/samplers/hmc.jl:150
 [38] sample(::Function, ::HMC{Any}) at /Users/hg344/.julia/dev/Turing/src/samplers/hmc.jl:106
 [39] top-level scope at none:0
```

### PR DESCRIPTION
**Pull request by willtebbutt**
Resolves #550 . The issue was indeed to do with the sharing of mutable state. The fix was simply to ensure that indexing gives copies of `vi.vals` rather than views into `vi.vals`.

### CODE REVIEW

File: src/core/varinfo.jl [hunk 1]
```diff
@@ -153,64 +153,32 @@ syms(vi::VarInfo) = map(vn -> vn.sym, vns(vi))  # get all symbols
...
-  #     r
-  # else
-    istrans(vi, vns[1]) ?
-      invlink(dist, reconstruct(dist, getval(vi, vns), length(vns))) :
-      reconstruct(dist, getval(vi, vns), length(vns))
-  # end
+function Base.getindex(vi::VarInfo, vns::Vector{VarName})
+    @assert haskey(vi, vns[1]) "[Turing] attempted to replay unexisting variables in VarInfo"
+    dist = getdist(vi, vns[1])
+    return copy(istrans(vi, vns[1]) ?
```
Comment (yebai) [inline src/core/varinfo.jl]: This looks a bit strange. Do you want to:

- `copy(istrans(vi, vns[1])` or
- `copy(invlink(dist, reconstruct(dist, getval(vi, vns), length(vns))))` and `copy( reconstruct(dist, getval(vi, vns), length(vns))))`

**Comment by trappmartin** [comment]
Maybe it would be better to change the behaviour of `getval` and `getall` instead?

File: src/core/varinfo.jl (hunk 1 above)
Comment (yebai) [reply src/core/varinfo.jl]: I see. Might be helpful to introduce an additional variable here for readability 

```julia

res = istrans(vi, vns[1]) ?
        invlink(dist, reconstruct(dist, getval(vi, vns), length(vns))) :
        reconstruct(dist, getval(vi, vns), length(vns))
return res
```

**Comment by mohamed82008** [comment]
I would just copy on use instead of changing the behavior of `getindex`. If the underlying data type of an indexable is mutable or has mutable fields at any level, then a shared state is the right mental model IMO. If this is a problem, then the following seem like reasonable fixes to me: 
1) Make the underlying type `isbits`, i.e. all fields and sub-fields are immutable,
2) Make a `copy` or `deepcopy` at call site after indexing, or
3) Make a convenience function that copies on indexing and call that instead.

**Comment by willtebbutt** [comment]
I agree with you in principle @mohamed82008, but in this case I believe our indexing behaviour should essentially behave the same as indexing into an array, so copying is appropriate. We should probably support a separate `view` operation to support a non-copying version.

**Comment by trappmartin** [comment]
@willtebbutt can I help with something to fix this PR? It's breaking quite a lot at the moment and I'm happy to help if possible.

**Comment by willtebbutt** [comment]
Apologies for the delay on this. I've been out of action for the last 7 days due to illness. I'm back in the office for the first time today and am working on this issue currently. I'll give you a shout if I get stuck.

**Comment by willtebbutt** [comment]
@yebai @xukai92 is there any particular reason that we have `getval`, `getvals`, and `getall`? ctrl-f-ing through the code base indicates that we don't use them other than to implement the internals of `getindex` for `VarInfo`.

**Comment by yebai** [comment]
@willtebbutt Functions like `getval`, `getvals` and `getall` are probably not necessary. I don't think they are used anywhere else apart from inside the `VarInfo` module.

**Pull request merged**